agent/grpc-middleware/rate.go

```diff
@@ -0,0 +1,55 @@
+package middleware
+
+import (
+	"context"
+	"errors"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/tap"
+
+	"github.com/hashicorp/consul/agent/consul/rate"
+)
+
+// ServerRateLimiterMiddleware implements a ServerInHandle function to perform
+// RPC rate limiting at the cheapest possible point (before the full request has
+// been decoded).
+func ServerRateLimiterMiddleware(limiter RateLimiter) tap.ServerInHandle {
+	return func(ctx context.Context, info *tap.Info) (context.Context, error) {
+		peer, ok := peer.FromContext(ctx)
+		if !ok {
+			// This should never happen!
+			return ctx, status.Error(codes.Internal, "gRPC rate limit middleware unable to read peer")
+		}
+
+		err := limiter.Allow(rate.Operation{
```

Comment:
Just realised that we should probably special-case the xDS endpoint because it implements its own concurrency limiting.

Reply:
Not sure I understand, the xDS concurrency limiter is quiet different,isn'it? 🤔

Reply:
Yeah, I'd say we should just _not_ rate-limit the xDS endpoint here, and trust the xDS server to do its own limiting.

Particularly because xDS sessions are long-lived and touch many different resources (so what kind of limit would we even apply here?)

Reply:
I think this is reasonable. In all cases the xDS endpoint would end up using data from watches which perform RPC calls, so I guess that would be enough to protect the server.